### DISCUSSION TITLE
Reusing base HTTP handlers for optimization

### DISCUSSION THREAD

**Discussion by JairoCA** [netty/netty#11785] [Q&A]
Hello,

I've been performing some profiling test to discover that 5% of the CPU time spent on the `nioEventLoopGroup` threads is spent on the `ServerInitializer.initChannel`.

I wonder if there is a way to clean and reuse the full pipeline after the channel is inactive using a pool of pipelines and in case that is not possible, I wonder if it would be worthy to add some sort of `clean()` method to `HttpRequestDecoder`, `HttpObjectAggregator` or `HttpResponseEncoder` so that it is not needed to create a new instance of them for each new client.

PS: obviously, this would use some sort of pooling to keep references of the previous pipelines / handlers

Thanks!

**Comment by NiteshKant** [comment]
Please share more details around what in `initChannel()` is consuming CPU. In my experience, the overhead is of propagating events (added, registered, active) on the pipeline and not the pipeline object creation itself. If this is what you see, then pooling the pipeline instance would not be that useful.

**Reply by JairoCA** [reply]
Thanks for your answer.

In the trace I'm attaching you can find 9% usage of `initChannel()` out of 17409 invocations (17409 requests processed). I have marked the constructors of the 3 handlers in the pipeline with a red line. I have a 4th custom handler, but that one is `Sharable`.

![trace](https://user-images.githubusercontent.com/7058733/138644428-8f3a1b84-cd1e-4023-b454-c7716364e512.png)

Additionally, in the next image you can see a better separation of the `addLast()`, that takes 6% of CPU.

![addLast](https://user-images.githubusercontent.com/7058733/138644946-3b6e3c81-30d2-4164-8a1e-a325e743d870.png)

Thanks!

**Reply by NiteshKant** [reply]
Thanks for that info, this does look like overhead of pipeline creation. 

> 17409 invocations (17409 requests processed)

Before trying pooling pipelines, have you tried pooling connections on the client side? Connection per request is usually costly even outside your application, specially if you use a secure protocol.

**Reply by JairoCA** [reply]
Thanks for your answer.

Sorry, I think I didn't explain myself well. The 17409 invocations belong to the server-side processing. The client-side is simulating 17409 different users from a different process in a different machine.

While it is true that I'm not using HTTP/1.1 reusing connections for my test, I think the test still shows some code sections that could be improved and affect every new user opening a connection with the server.

**Reply by NiteshKant** [reply]
Not disagreeing that there are code sections that can be improved, it is a trade-off about more complex API contract for accommodating reuse VS the benefit it provides. Since, new connection establishment overhead is usually ignorable when reusing connections, I feel introducing this complexity may not be desirable. However, that is just my view, I will wait for others to chime in.

**Reply by JairoCA** [reply]
Thanks for your answer.

I think you are right that if we reuse connections, most of the problem will be gone. However, according to my architecture and distribution of users between servers, it may not happen often that users send sequences of requests to the same servers.

I will think about the situation a bit more and try to implement the reuse of connections, as I don't think it will waste much resources anyway.

Thanks again :)

**Comment by doom369** [comment]
@JairoCA There are 2 simple things you can do to improve the CPU usage for `initCahnnel()` :

- Use a predefined name for your handlers:
 `.addLast("HttpServerCodec", new HttpServerCodec())` instead of `.addLast(new HttpServerCodec())`
- Override `boolean isSharable()` method in your `ChannelHandlerAdapter` handlers and make it no-op (after you're make sure your pipeline is correct)

These changes are super simple and will save you ~30% for `initCahnnel()`.

**Reply by JairoCA** [reply]
Wow! These hints worked great and exactly as you predicted around 30% improvement.

Do you think there might be more hints like these that I may be missing?

Thank you very much,
Jairo

**Reply by doom369** [reply]
Also, you may try to override `ChannelPipeline` and remove all unnecessary logic. The only other option is to fork the netty and remove the old "legacy" + "infrastructure" code, as many things are hidden by private methods/fields.
